plugins/importexport/native/filter/SubmissionFileNativeXmlFilter.php

```diff
@@ -148,48 +146,63 @@ public function createSubmissionFileNode($doc, $submissionFile)
         $revisions = Repo::submissionFile()->getRevisions(($submissionFile->getId()));
+        $basePath = rtrim(Config::getVar('files', 'files_dir'), '/') . '/';
+        $hasRevision = false;
         foreach ($revisions as $revision) {
-            $localPath = rtrim(Config::getVar('files', 'files_dir'), '/') . '/' . $revision->path;
+            $localPath = $basePath . $revision->path;
+            if (!file_exists($localPath)) {
+                $deployment->addWarning(PKPApplication::ASSOC_TYPE_SUBMISSION_FILE, $submissionFile->getId(), __('plugins.importexport.native.error.submissionFileRevisionMissing', ['id' => $submissionFile->getId(), 'revision' => $revision->revision_id, 'path' => $localPath]));
+                continue;
+            }
+            $hasRevision = true;
+
             $revisionNode = $doc->createElementNS($deployment->getNamespace(), 'file');
             $revisionNode->setAttribute('id', $revision->fileId);
             $revisionNode->setAttribute('filesize', filesize($localPath));
             $revisionNode->setAttribute('extension', pathinfo($revision->path, PATHINFO_EXTENSION));
+            $submissionFileNode->appendChild($revisionNode);
 
-            if (array_key_exists('no-embed', $this->opts)) {
-                $hrefNode = $doc->createElementNS($deployment->getNamespace(), 'href');
-                if (array_key_exists('use-file-urls', $this->opts)) {
-                    $stageId = Repo::submissionFile()->getWorkflowStageId($submissionFile);
-                    $dispatcher = Application::get()->getDispatcher();
-                    $request = Application::get()->getRequest();
-                    $params = [
-                        'submissionFileId' => $submissionFile->getId(),
-                        'submissionId' => $submissionFile->getData('submissionId'),
-                        'stageId' => $stageId,
-                    ];
-                    $url = $dispatcher->url($request, PKPApplication::ROUTE_COMPONENT, $context->getPath(), 'api.file.FileApiHandler', 'downloadFile', null, $params);
-                    $hrefNode->setAttribute('src', $url);
-                } else {
-                    $hrefNode->setAttribute('src', $revision->path);
-                }
-                $hrefNode->setAttribute('mime_type', $revision->mimetype);
-                $revisionNode->appendChild($hrefNode);
-            } else {
+            if (!($this->opts['no-embed'] ?? false)) {
                 $embedNode = $doc->createElementNS($deployment->getNamespace(), 'embed', base64_encode(file_get_contents($localPath)));
                 $embedNode->setAttribute('encoding', 'base64');
                 $revisionNode->appendChild($embedNode);
+                continue;
             }
 
-            $submissionFileNode->appendChild($revisionNode);
+            $hrefNode = $doc->createElementNS($deployment->getNamespace(), 'href');
+            $revisionNode->appendChild($hrefNode);
```

Comment:
It does not change the result, but I would prefer the `$revisionNode->appendChild($hrefNode);` to came after all the attributes and children of `$hrefNode` have been added, if possible - `$revisionNode->appendChild($hrefNode);` being something like the "closure statement" of a process. What do you say?

Reply:
I also prefer to build the tree sequentially.
I just moved this line up to "return early" without duplicating the `appendChild()`, I'll review how many early returns I added, if it's few, then it's ok to duplicate 👌